src/psm/src/ir_solver.h

```diff
@@ -48,207 +57,190 @@ class Resizer;
-  std::vector<std::pair<odb::dbInst*, float>> getPower();
-
-  bool getConnectionTest() const;
-  int getMinimumResolution() const;
-
-  void writeSpiceFile(const std::string& file) const;
-  void writeVoltageFile(const std::string& file) const;
-  void writeEMFile(const std::string& file) const;
-  void writeErrorFile(const std::string& file) const;
-
-  bool build(const std::string& error_file = "",
-             bool connectivity_only = false);
-
-  const std::vector<SourceData>& getSources() const { return sources_; }
-
-  double getWorstCaseVoltage() const { return wc_voltage_; }
-  double getMaxCurrent() const { return max_cur_; }
-  double getAvgCurrent() const { return avg_cur_; }
-  int getNumResistors() const { return num_res_; }
-  double getAvgVoltage() const { return avg_voltage_; }
-  float getSupplyVoltageSrc() const
-  {
-    if (supply_voltage_src_.has_value()) {
-      return supply_voltage_src_.value();
-    } else {
-      return 0.0;
-    }
-  }
+           const std::map<odb::dbNet*, std::map<sta::Corner*, Voltage>>&
+               user_voltages,
+           const PDNSim::GeneratedSourceSettings& generated_source_settings);
+
+  odb::dbNet* getNet() const { return net_; };
+
+  bool check();
+
+  void solve(sta::Corner* corner,
+             GeneratedSourceType source_type,
+             const std::string& source_file);
+
+  void report(sta::Corner* corner) const;
+  void reportEM(sta::Corner* corner) const;
+
+  Results getSolution(sta::Corner* corner) const;
+  EMResults getEMSolution(sta::Corner* corner) const;
+  PDNSim::IRDropByPoint getIRDrop(odb::dbTechLayer* layer,
+                                  sta::Corner* corner) const;
+  ConnectivityResults getConnectivityResults() const;
+
+  void enableGui(bool enable);
+
+  void writeErrorFile(const std::string& error_file) const;
+  void writeInstanceVoltageFile(const std::string& voltage_file,
+                                sta::Corner* corner) const;
+  void writeEMFile(const std::string& em_file, sta::Corner* corner) const;
+  void writeSpiceFile(GeneratedSourceType source_type,
+                      const std::string& spice_file,
+                      sta::Corner* corner,
+                      const std::string& voltage_source_file) const;
+
+  bool belongsTo(Node* node) const;
+  bool belongsTo(Connection* connection) const;
+
+  std::vector<sta::Corner*> getCorners() const;
+  bool hasSolution(sta::Corner* corner) const;
+  Voltage getNetVoltage(sta::Corner* corner) const;
+  std::optional<Voltage> getVoltage(sta::Corner* corner, Node* node) const;
+
+  std::pair<bool, Voltage> getSDCVoltage(sta::Corner* corner,
+                                         odb::dbNet* net) const;
+  std::pair<bool, Voltage> getPVTVoltage(sta::Corner* corner) const;
+  std::pair<bool, Voltage> getUserVoltage(sta::Corner* corner,
+                                          odb::dbNet* net) const;
+  std::pair<bool, Voltage> getSolutionVoltage(sta::Corner* corner) const;
```

Comment:
I believe the bool here is for success.  Perhaps std::optional might be clearer.

Reply:
good, point, I forget the std::optional is available